bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -227,28 +224,14 @@ func (r *KubeadmConfigReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		return ctrl.Result{}, nil
 	// Status is ready means a config has been generated.
 	case config.Status.Ready:
-		// If the BootstrapToken has been generated for a join and the infrastructure is not ready.
-		// This indicates the token in the join config has not been consumed and it may need a refresh.
 		if (config.Spec.JoinConfiguration != nil && config.Spec.JoinConfiguration.Discovery.BootstrapToken != nil) && !configOwner.IsInfrastructureReady() {
-
-			token := config.Spec.JoinConfiguration.Discovery.BootstrapToken.Token
-
-			remoteClient, err := r.remoteClientGetter(ctx, r.Client, util.ObjectKey(cluster))
-			if err != nil {
-				log.Error(err, "Error creating remote cluster client")
-				return ctrl.Result{}, err
-			}
-
-			log.Info("Refreshing token until the infrastructure has a chance to consume it")
-			err = refreshToken(ctx, remoteClient, token)
-			if err != nil {
-				// It would be nice to re-create the bootstrap token if the error was "not found", but we have no way to update the Machine's bootstrap data
-				return ctrl.Result{}, errors.Wrapf(err, "failed to refresh bootstrap token")
-			}
-			// NB: this may not be sufficient to keep the token live if we don't see it before it expires, but when we generate a config we will set the status to "ready" which should generate an update event
-			return ctrl.Result{
-				RequeueAfter: DefaultTokenTTL / 2,
-			}, nil
+			// If the BootstrapToken has been generated for a join and the infrastructure is not ready.
+			// This indicates the token in the join config has not been consumed and it may need a refresh.
+			return r.refreshBootstrapToken(ctx, config, cluster)
+		} else if (config.Spec.JoinConfiguration != nil && config.Spec.JoinConfiguration.Discovery.BootstrapToken != nil) && configOwner.IsMachinePool() {
+			// If the BootstrapToken has been generated and infrastructure is ready but the configOwner is a MachinePool,
+			// we rotate the token to keep it fresh for future scale ups.
+			return r.rotateMachinePoolBootstrapToken(ctx, config, cluster, scope)
```

Comment:
Background question.
If I got it right, in the case of machine pools each token expire after TTL, being replaced by a new one. What happens to a joining machine which starts joining with a token but then for any reason (e.g poor network performances) takes longer than TTL before actually running kubeadm join? is it expected to fail?

Should we make this a little bit smarter and refresh in case a machine pool is scaling, rotate only if the machine pool is idle?

Reply:
> What happens to a joining machine which starts joining with a token but then for any reason (e.g poor network performances) takes longer than TTL before actually running kubeadm join?

If the Machine takes longer than a TTL to join it'd fail regardless if it was MachinePool or not, that could also happen in the case of simple machines, or the ones in a MachineDeployment.

> Should we make this a little bit smarter and refresh in case a machine pool is scaling, rotate only if the machine pool is idle?

IIRC, this is already the case and it's covered by `refreshBootstrapToken`. @CecileRobertMichon correct me if I'm wrong, the assumption I was making is that MachinePool flips between Ready/NotReady while it's scaling up/down.

Reply:
> If the Machine takes longer than a TTL to join it'd fail regardless if it was MachinePool or not, that could also happen in the case of simple machines, or the ones in a MachineDeployment.

For single machines the token TTL is extended until Machine.infrastructureReady is true, so a single machine won't fail in case of slow startup

> IIRC, this is already the case and it's covered by refreshBootstrapToken

Uhmm might be I'm missing something, but I can't find where we are testing the machine pool status

Reply:
> IIRC, this is already the case and it's covered by refreshBootstrapToken. @CecileRobertMichon Cecile Robert-Michon FTE correct me if I'm wrong, the assumption I was making is that MachinePool flips between Ready/NotReady while it's scaling up/down.

Correct, regardless of the infrastructure type, if `!configOwner.IsInfrastructureReady()` it would go into the case above to refresh the token. The token is only rotated if the MachinePool infra is ready. This is being tested here: https://github.com/kubernetes-sigs/cluster-api/pull/3940/files#diff-ad5b3c5b9e092b1dced2a9f66a655d659462a11bd6138e7c3f9853723f924f9aR976